### ISSUE TITLE
How can I manually tell the TileLayerOptions to load and paint a tile?

### ISSUE THREAD

**Issue by JaffaKetchup** [fleaflet/flutter_map#966]
### Discussed in https://github.com/fleaflet/flutter_map/discussions/963

<div type='discussions-op-text'>

<sup>Originally posted by **JaffaKetchup** July  5, 2021</sup>
I need to do this as I'm implementing a way to remove grey tiles, but I can't change the base code in `flutter_map`. See https://github.com/JaffaKetchup/flutter_map_tile_caching/issues/12.

My existing code:
```dart
@override
ImageProvider getImage(Coords<num> coords, TileLayerOptions options) {
    // `preloadSurroundings` is a boolean dictating whether the grey tiles should be attempted to be removed
    if (preloadSurroundings) {
      for (double x = coords.x - 6; x < coords.x + 6; x++) {
        for (double y = coords.y - 6; y < coords.y + 6; y++) {
          /*_CachedTileImageProvider(
            getTileUrl(Coords<num>(x, y)..z = coords.z, options),
            Coords<num>(coords.x, coords.y)..z = coords.z,
            cacheName: cacheName,
          );*/
         // How can I manually load AND PAINT the image above
        }
      }
    }
    return _CachedTileImageProvider(
      getTileUrl(coords, options),
      Coords<num>(coords.x, coords.y)..z = coords.z,
      cacheName: cacheName,
    );
  }
```

**Comment by ibrierley** [comment]
Isn't this just the same issue as 963?

**Comment by JaffaKetchup** [comment]
Yeah, I can close it if you want (as duplicate), but it was to attract attention because no-one had answered, so I wasn't sure if anybody had actually seen it, or no-body knew. That's why it has a 'Discussed in' header.

**Comment by ibrierley** [comment]
The issue also confuses me, as typically grey tiles are just transparent tiles that aren't available yet aren't they...so what's actually trying to be achieved ?

**Comment by JaffaKetchup** [comment]
Yeah, sorry, should've clarified. This is for my plugin that handles caching.

A user wanted to preload a whole area after it had been downloaded (https://github.com/JaffaKetchup/flutter_map_tile_caching/issues/12) to totally avoid the grey tiles (which still appeared even if the tile was cached/downloaded already because of the time to lookup, fetch from the caching database and paint).

I suggested that preloading a whole area would be unnecessary and could cause extreme performance issues (if 20000 tiles (currently max number of cached tiles allowed) were shown, the map would be incredibly laggy), so the idea is to automatically get the nearest tiles in each direction before the user scrolls there.

Therefore I need to paint a tile before it actually becomes visible, the topic of this issue/discussion.

**Comment by ibrierley** [comment]
I've certainly played with preloading tiles (iirc I preloaded an extra tile the user was panning to). But I had to edit the tile layer code myself around the tileRange.min.y for loop (which looks partially what you are trying already)...Only thing I can think of is to have your own tile_layer code (or have it as part of the plugin), but maybe that's an issue.

Just thinking though (taking a quick scan of the plugin), do you need to use a database, why not just use a file cache ? (just wondering if that's faster to get around the problem another way).

**Comment by JaffaKetchup** [comment]
Yeah, I hoped I wouldn't have to edit the code in this package.
The main reason I can't use files is that the plugin was originally created by someone else using sqflite, so instead of changing it completely, I've just stuck with it. It also allows better queries etc..

**Comment by JaffaKetchup** [comment]
Actually, I've reconsidered. Perhaps using a file method would be better. I can still find some code from one of my previous attempts at https://github.com/fleaflet/flutter_map/pull/798, so @ibrierley, do you think using a file method would be quicker?

**Comment by ibrierley** [comment]
I suspect it "may" be :D, I've been using   await
DefaultCacheManager().putFile()    await
DefaultCacheManager().getFileFromCache() with
https://pub.dev/packages/flutter_cache_manager as its felt relatively
straightforward to implement (there's also await
DefaultCacheManager().getSingleFile(url) which I'm using);

**Comment by JaffaKetchup** [comment]
[That package](https://pub.dev/packages/flutter_cache_manager) looks good at first glimpse, but this worries me:

> By default the cached files are stored in the temporary directory of the app. This means the OS can delete the files any time.
> 
> Information about the files is stored in a database using sqflite on Android, iOS and macOs, or in a plain JSON file on other platforms. The file name of the database is the key of the cacheManager, that's why that has to be unique.

I need to not store in app cache, because the OS deleting map tiles could have bad consequences (extreme example: hiker downloads an area, without the hiker knowing, the OS deletes it, the hiker gets stuck on a mountain with no signal). I also need to have multiple caching spaces to keep a functionality I recently added. This may be solvable if I specify where else to put it?

And surely by using a database, it won't make a difference to the time again.

Although this package has nice features built-in like having an expiry, I may have to use [network_to_file_image](https://pub.dev/packages/network_to_file_image), and make that expiry functionality another way using metadata of some sorts?

But yeah, I probably will have a go with this, to see if it is indeed faster. Hopefully, it'll cut down on my library's size as well (less code). I'll try the package you mentioned first, but I feel I may need to use that other library instead. Thanks for the suggestion!


However, in the event that neither of these two options is faster, I'll still need to do the original purpose of this issue, so I'm leaving this issue open for that.

**Comment by JaffaKetchup** [comment]
I've done the basics required, and unfortunately, I can't see that much of a difference. I'll still have to do this, but a way might be to make the map bigger than normal, so it overflows the visible area?

**Comment by JaffaKetchup** [comment]
Ok, so that doesn't work very well either.

So the question still stands, 'How can I manually tell the TileLayerOptions to load and paint a tile?'

**Comment by JaffaKetchup** [comment]
Hi there,
Does anyone else have any ideas, or I'll close this for now?
Thanks.

**Comment by ibrierley** [comment]
Tbh, given the sluggish turnaround of changes here anyway, I'd probably go with a custom plugin alternate tile provider.

**Comment by JaffaKetchup** [comment]
>given the sluggish turnaround of changes here

😂

Probably misunderstanding here, but does that ('custom plugin alternate tile provider') mean it can be done without submitting a PR for this library?

**Comment by ibrierley** [comment]
Yes, I just create a map plugin, and don't use the default tile layer any
more (I find that too hard to follow these days anyway). For example...

https://github.com/ibrierley/flutter_map_vector_tiles/blob/master/lib/VectorTileWidget.dart

Eg line 81 & 99 defines the map plugin, and I basically loop over the tiles
needed around line 244, and grab the tiles around line 438, which could be
grabbing from a specific saved region of tiles for example. My stuff is too
messy, but it may give an outline of what I'm doing iyswim.

**Comment by ibrierley** [comment]
Just to add, the downside is you end up with some duplicated tile bits of
code you will see there (eg _isValidTile() method) ..there may be a way to
extend the tile layer using a plugin, but kind of got a bit lost with
trying that approach.

**Comment by ibrierley** [comment]
Sorry for the spam, in fact I'd almost be tempted to just make a pure
duplicate of the flutter_map tile layer as a map plugin (I don't think too
much needs changing to do that, if you look at my example of creating the
map plugin), edit as you need, may be simpler, and maybe others could use
it then as well. I know it would be 99% duplicated code, but it may allow
you/others to change as needed faster.

**Comment by JaffaKetchup** [comment]
I see. So I'm assuming thats using the Plugin API (plugins: [])?

**Comment by JaffaKetchup** [comment]
Don't worry, I just answered my own question looking at line 84 of your vector plugin.

Thanks again, and I'll see what I can reasonably do with a tile layer (ie. As long as it doesn't triple the size of my library)!

**Comment by JaffaKetchup** [comment]
For now I'm going to close this, even though I haven't yet done what @ibrierley suggested - I found it quite complicated. Thanks for the help!
